File: 프로그래머스/괄호 변환.cpp
#include <string>
#include <stack>
#include <vector>
#include <iostream>

using namespace std;

int check(string s)
{
    stack<char> st;
    for(int i=0; i<s.size(); i++){
        if(s[i]=='('){
            st.push(s[i]);
        }
        else{
            if(st.empty()){
                return (0);
            }
            else{
                st.pop();
            }
        }
    }
    if(!st.empty()){
        return 0;
    }
    return 1;
}




string cal(string p)
{
    if(p.size()==0){
        return p;
    }
    string u;
    string v;
    int r=0;
    int l=0;
    for(int i=0; i<p.size(); i++){
        if(p[i]=='('){
            l++;
        }
        else{
            r++;
        }
        if(r==l){
            u=p.substr(0,i+1);
            break;
        }
    }
    v = p.substr(u.size());
    if (check(u))
    {
        return u+=cal(v);
    }
    else
    {
        string temp="(";
        temp += cal(v);
        temp += ")";
        for(int i=1; i<u.size()-1; i++){
            if(u[i]==')')
                temp+='(';
            else{
                temp+=')';
            }
        }
        return temp;
    }
}

string solution(string p)
{
    string answer = "";
    answer = cal(p);
    return answer;
}

File: 프로그래머스/괄호 변환.py
def check(s):
    li = []
    for i in s:
        if(i=='('):
            li.append(i)
        else:
            if(li!=[]):
                li.pop();
            else:
                return False;
    if(li!=[]):
        return False;
    return True

def cal(s):
    if(s==""):
        return ""
    cnt1 = 0
    cnt2 = 0
    u = ""
    for i in s:
        u+=i
        if(i=='('):
            cnt1+=1
        else:
            cnt2+=1
        if(cnt1==cnt2):
            break
    v = s[len(u):]
    if(check(u)):
        return u + cal(v)
    else:
        temp = '('
        temp += cal(v)
        temp += ')'
        for j in u[1:len(u)-1]:
            if(j=='('):
                temp+=')'
            else:
                temp+='('
        return temp

def solution(p):
    answer = cal(p)
    return answer
